### PR TITLE
[HiDream LoRA] optimizations + small updates

### PR DESCRIPTION
**Pull request by linoytsaban**
some memory optimizations: 

- [x] add pre-computation of prompt embeddings when custom prompts are used as well
- [x] add pre-computation of validation prompt as well
- [x] add `--skip_final_inference` - to allow to run with validation, but skip the final loading of the pipeline with the lora weights to reduce memory reqs

other changes:

- [x] update default trained layers
- [x] save model card even if model is not pushed to hub
- [x] remove scheduler initialization from code example - not necessary anymore (it's now if the base model's config)

todo: 

- [x] update readme with better defaults 

Yarn Art LoRA
Screenshot 2025-04-23 at 14 14 22

<details>
<summary>training config</summary>

```
import os
os.environ['MODEL_NAME'] = "HiDream-ai/HiDream-I1-Full"
os.environ['DATASET_NAME'] ="Norod78/Yarn-art-style"
os.environ['OUTPUT_DIR'] = "hidream-yarn-art-lora-v2-trainer"

!accelerate launch train_dreambooth_lora_hidream.py \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --dataset_name=$DATASET_NAME \
  --output_dir=$OUTPUT_DIR \
  --mixed_precision="bf16" \
  --lora_layers="to_k,to_q,to_v,to_out"\
  --instance_prompt="a dog, yarn art style" \
  --validation_prompt="yoda, yarn art style" \
  --caption_column="text" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 \
  --use_8bit_adam\
  --rank=16 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant_with_warmup" \
  --lr_warmup_steps=200 \
  --max_train_steps=1000 \
  --validation_epochs=25 \
  --seed="0" \
  --push_to_hub
```

### CODE REVIEW

**Comment by linoytsaban** [comment]
@sayakpaul now that I'm thinking about it, even when validation is enabled, since we're not optimizing the text encoders,  can't we just pre-encode the validation prompt embeddings as well? and then we don't need to keep or load text encoders for validation at all and simply pass the embeddings to `log_validation`

**Comment by sayakpaul** [comment]
> @sayakpaul now that I'm thinking about it, even when validation is enabled, since we're not optimizing the text encoders, can't we just pre-encode the validation prompt embeddings as well? and then we don't need to keep or load text encoders for validation at all and simply pass the embeddings to `log_validation`

We should. Let's do that.

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 1]
```diff
@@ -253,8 +248,7 @@ def log_validation(
             )
 
     del pipeline
-    if torch.cuda.is_available():
-        torch.cuda.empty_cache()
+    free_memory()
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:251]: If we're doing pre-encoding of prompts, then it is important to first move the model components of the `pipeline` (the `pipeline` object for doing prompt encoding) to CPU and then perform the rest of the stuff for memory savings. 

Reference:
https://github.com/huggingface/diffusers/issues/11376#issuecomment-2820351451

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 2]
```diff
@@ -392,6 +386,15 @@ def parse_args(input_args=None):
...
         help="A prompt that is used during validation to verify that the model is learning.",
     )
+
+    parser.add_argument(
+        "--skip_final_inference",
+        default=False,
+        action="store_true",
+        help="whether to skip the final inference step with loaded lora weights upon training completion. "
+             "Set to True to reduce memory",
+    )
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:396]:
```suggestion
    parser.add_argument(
        "--skip_final_inference",
        default=False,
        action="store_true",
        help="Whether to skip the final inference step with loaded lora weights upon training completion. This will run intermediate validation inference if `validation_prompt` is provided. Specify to reduce memory.",
    )
```

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 3]
```diff
@@ -1367,20 +1362,39 @@ def compute_text_embeddings(prompt, text_encoding_pipeline):
 
     vae_config_scaling_factor = vae.config.scaling_factor
     vae_config_shift_factor = vae.config.shift_factor
-    if args.cache_latents:
+
+    # if cache_latents is set to True, we encode images to latents and store them.
+    # Similarly to the pre-encoding in the case of a single instance prompt, if custom prompts are provided
+    # we encode them in advance as well (by iterating over training batches)
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1368]: Slightly less confusing I believe?

```suggestion
    # Similar to pre-encoding in the case of a single instance prompt, if custom prompts are provided
    # we encode them in advance as well.
```

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 4]
```diff
@@ -1367,20 +1362,39 @@ def compute_text_embeddings(prompt, text_encoding_pipeline):
...
+    precompute_latents = args.cache_latents or train_dataset.custom_instance_prompts
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1369]: Can't we compute it after initializing the dataloader similar to how the latents are precomputed??

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 5]
```diff
@@ -1367,20 +1362,39 @@ def compute_text_embeddings(prompt, text_encoding_pipeline):
...
+                    t5_prompt_cache.append(t5_prompt_embeds)
+                    llama3_prompt_cache.append(llama3_prompt_embeds)
+                    pooled_prompt_cache.append(pooled_prompt_embeds)
 
+            # delete tokenizers and text ecnoders except for llama (tokenizer & te four)
+            # as it's needed for inference with pipeline
         if args.validation_prompt is None:
-            del vae
-            free_memory()
+            del vae, tokenizer_four, text_encoder_four
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1395]: Important to note:
https://github.com/huggingface/diffusers/issues/11376#issue-3008144624

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 6]
```diff
@@ -1485,11 +1499,19 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
+                #     llama3_prompt_embeds = llama3_prompt_cache[step]
+                #     pooled_prompt_embeds = pooled_prompt_cache[step]
                 # encode batch prompts when custom prompts are provided for each image -
                 if train_dataset.custom_instance_prompts:
-                    t5_prompt_embeds, llama3_prompt_embeds, pooled_prompt_embeds = compute_text_embeddings(
-                        prompts, text_encoding_pipeline
-                    )
+                    # t5_prompt_embeds, llama3_prompt_embeds, pooled_prompt_embeds = compute_text_embeddings(
+                    #     prompts, text_encoding_pipeline
+                    # )
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1511]: Remove?

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 7]
```diff
@@ -1655,42 +1679,43 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
-            args.num_validation_images = args.num_validation_images if args.num_validation_images else 1
-            pipeline_args = {"prompt": prompt_to_use, "num_images_per_prompt": args.num_validation_images}
-            images = log_validation(
-                pipeline=pipeline,
-                args=args,
-                accelerator=accelerator,
-                pipeline_args=pipeline_args,
-                epoch=epoch,
-                is_final_validation=True,
+        if not args.skip_final_inference:
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1683]:
```suggestion
        should_run_final_inference = (not args.skip_final_inference) or (args.validation_prompt and args.num_validation_images > 0) or (args.final_validation_prompt)
        if should_run_final_inference:
```

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 8]
```diff
@@ -1655,42 +1679,43 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
+                text_encoder_4=text_encoder_4,
+                revision=args.revision,
+                variant=args.variant,
                 torch_dtype=weight_dtype,
             )
+            # load attention processors
+            pipeline.load_lora_weights(args.output_dir)
+
+            # run inference
+            if (args.validation_prompt and args.num_validation_images > 0) or (args.final_validation_prompt):
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1706]: We can safely remove with the previous suggestion.

**Review by sayakpaul — COMMENTED** [review]
Thanks for the optims. LMK if the comments make sense or if anything is unclear.

File: examples/dreambooth/train_dreambooth_lora_hidream.py (hunk 4 above)
Comment (linoytsaban) [reply examples/dreambooth/train_dreambooth_lora_hidream.py:1366]: not sure what I follow, it is currently done after initializing the data-loader, simultaneously with the pre-computation of the latents as you say

**Comment by linoytsaban** [comment]
@bot /style

**Comment by linoytsaban** [comment]
@bot /style

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 9]
```diff
@@ -1140,7 +1131,7 @@ def main(args):
     if args.lora_layers is not None:
         target_modules = [layer.strip() for layer in args.lora_layers.split(",")]
     else:
-        target_modules = ["to_k", "to_q", "to_v", "to_out.0"]
+        target_modules = ["to_k", "to_q", "to_v", "to_out", "to_out_t", "to_q_t", "to_v_t", "to_k_t"]
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1134]: Oh interesting. What changed here?

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 10]
```diff
@@ -1367,20 +1363,65 @@ def compute_text_embeddings(prompt, text_encoding_pipeline):
...
+            ) = compute_text_embeddings(args.validation_prompt, text_encoding_pipeline)
+        # move back to cpu before deleting to ensure memory is freed see: https://github.com/huggingface/diffusers/issues/11376#issue-3008144624
+        if args.offload or args.cache_latents:
+            vae = vae.to("cpu")
+            if args.cache_latents:
+                del vae
+
+    # move back to cpu before deleting to ensure memory is freed see: https://github.com/huggingface/diffusers/issues/11376#issue-3008144624
+    text_encoding_pipeline = text_encoding_pipeline.to("cpu")
+    print(text_encoder_one.device)
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1413]: Remove?

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 11]
```diff
@@ -1619,26 +1660,24 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
                 # create pipeline
                 pipeline = HiDreamImagePipeline.from_pretrained(
                     args.pretrained_model_name_or_path,
-                    tokenizer_4=tokenizer_four,
-                    text_encoder_4=text_encoder_four,
+                    tokenizer_4=None,
+                    text_encoder_4=None,
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1664]: Should we not specify the other tokenizers and text encoders to `None` as well?

File: examples/dreambooth/train_dreambooth_lora_hidream.py (hunk 8 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_hidream.py:1718]: I think this wasn't addressed?

**Review by sayakpaul — COMMENTED** [review]
Thank you! Left some more comments. I think we can merge this today. Also would be good to see if we test this on a 40GB GPU.

**Comment by linoytsaban** [comment]
Thanks @sayakpaul!

I added prints using your snippet ([here](https://github.com/huggingface/diffusers/issues/11376#issue-3008144624)) - right before caching & and pre-computation of prompt embeddings and straight after deleting them. 
freeing the memory seems to work as expected getting us to ~33GB, but we hit ~60GB prior when we move the text encoding pipeline to gpu 

***with offloading*** - 
```
=== CUDA Memory Stats before caching ===
Current allocated: 32.24 GB
Max allocated: 32.24 GB
Current reserved: 58.29 GB
Max reserved: 58.29 GB

=== CUDA Memory Stats after caching ===
Current allocated: 58.41 GB
Max allocated: 58.41 GB
Current reserved: 60.84 GB
Max reserved: 60.84 GB

=== CUDA Memory Stats after freeing ===
Current allocated: 32.87 GB
Max allocated: 32.87 GB
Current reserved: 33.82 GB
Max reserved: 33.82 GB
```

***without offloading & caching*** -
```
=== CUDA Memory Stats before caching ===
Current allocated: 57.77 GB
Max allocated: 57.77 GB
Current reserved: 58.50 GB
Max reserved: 58.50 GB

=== CUDA Memory Stats after caching ===
Current allocated: 58.35 GB
Max allocated: 58.35 GB
Current reserved: 59.11 GB
Max reserved: 59.11 GB

=== CUDA Memory Stats after freeing ===
Current allocated: 32.98 GB
Max allocated: 32.98 GB
Current reserved: 33.39 GB
Max reserved: 33.39 GB
```

**Comment by linoytsaban** [comment]
@bot /style

**Comment by linoytsaban** [comment]
@bot /style

File: examples/dreambooth/train_dreambooth_lora_hidream.py [hunk 12]
```diff
@@ -1140,7 +1131,7 @@ def main(args):
     if args.lora_layers is not None:
         target_modules = [layer.strip() for layer in args.lora_layers.split(",")]
     else:
-        target_modules = ["to_k", "to_q", "to_v", "to_out.0"]
+        target_modules = ["to_k", "to_q", "to_v", "to_out"]
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_hidream.py:1134]: Perhaps we can add a comment explaining that including `to_out` will target all the expert layers.

**Review by sayakpaul — APPROVED** [review]
Looks great! Thanks Linoy!

**Comment by linoytsaban** [comment]
@bot /style
